### PR TITLE
[Tokenomics] Use service module directly to find a service

### PR DESCRIPTION
**Pull request by adshmh**
## Summary

The Tokenomics module uses the service module directly to find a service, in order to get the compute units per relay for the service.

## Issue

https://github.com/pokt-network/poktroll/blob/738e3429c06d7752cb3467b6b90c2710718700a7/x/tokenomics/keeper/settle_session_accounting.go#L233-L236

image

# https://github.com/pokt-network/poktroll/pull/706#issuecomment-2251467157

## Type of change

Select one or more:

- [ ] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

**Documentation changes** (only if making doc changes)
- [ ] `make docusaurus_start`; only needed if you make doc changes

**Local Testing** (only if making code changes)
- [x] **Unit Tests**: `make go_develop_and_test`
- [] **LocalNet E2E Tests**: `make test_e2e`
- See [quickstart guide](https://dev.poktroll.com/developer_guide/quickstart) for instructions

**PR Testing** (only if making code changes)
- [ ] **DevNet E2E Tests**: Add the `devnet-test-e2e` label to the PR.
    - **THIS IS VERY EXPENSIVE**, so only do it after all the reviews are complete.
    - Optionally run `make trigger_ci` if you want to re-trigger tests without any code changes
    - If tests fail, try re-running failed tests only using the GitHub UI as shown [here](https://github.com/pokt-network/poktroll/assets/1892194/607984e9-0615-4569-9452-4c730190c1d2)

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have commented my code
- [x] I have performed a self-review of my own code; both comments & source code
- [ ] I create and reference any new tickets, if applicable
- [ ] I have left TODOs throughout the codebase, if applicable

## Summary by CodeRabbit

- **New Features**
  - Introduced a new `ServiceKeeper` component, enhancing service management within various modules.
  - Added error handling for scenarios where a requested service cannot be located.

- **Bug Fixes**
  - Updated test cases to handle new service management features and ensure robust error handling.

- **Refactor**
  - Simplified service-related logic in the tokenomics module, improving code clarity and efficiency.

- **Tests**
  - Added new test cases to validate behavior regarding service retrieval and existence.

### CODE REVIEW

**Review by Olshansk — APPROVED** [review]
Small, clean and simple. Love it!

**Comment by Olshansk** [comment]
@adshmh I added the `devnet-test-e2e` tag. You need to run `make trigger_ci` to run it again.

File: x/tokenomics/keeper/keeper_settle_pending_claims_test.go [hunk 1]
```diff
@@ -83,29 +83,30 @@ func (s *TestSuite) SetupTest() {
 		preGeneratedAccts,
 	).String()
 
-	service := &sharedtypes.Service{Id: testServiceId}
+	service := sharedtypes.Service{Id: testServiceId}
+	s.keepers.SetService(s.ctx, service)
```
Comment (red-0ne) [inline x/tokenomics/keeper/keeper_settle_pending_claims_test.go:87]: It's maybe an opportunity to use `WithService` keeper opt function?

File: x/tokenomics/keeper/settle_session_accounting_test.go [hunk 2]
```diff
@@ -106,11 +107,6 @@ func TestSettleSessionAccounting_ValidAccounting(t *testing.T) {
 	app := apptypes.Application{
 		Address: sample.AccAddress(),
 		Stake:   &appStake,
```
Comment (red-0ne) [inline x/tokenomics/keeper/settle_session_accounting_test.go:109]: We should have `app` staked for `service`.

```suggestion
		Stake:   &appStake,
		ServiceConfigs: []*sharedtypes.ApplicationServiceConfig{
			{
				Service: service,
			},
		},
```

File: x/tokenomics/keeper/settle_session_accounting_test.go [hunk 3]
```diff
@@ -197,25 +201,13 @@ func TestSettleSessionAccounting_AppStakeTooLow(t *testing.T) {
...
-		OwnerAddress:         sample.AccAddress(),
-	}
 	// Add a new application
 	appStake := cosmostypes.NewCoin("upokt", math.NewInt(40000))
 	expectedAppEndStakeZeroAmount := cosmostypes.NewCoin("upokt", math.NewInt(0))
 	expectedAppBurn := appStake.AddAmount(math.NewInt(2000))
 	app := apptypes.Application{
 		Address: sample.AccAddress(),
 		Stake:   &appStake,
-		ServiceConfigs: []*sharedtypes.ApplicationServiceConfig{
```
Comment (red-0ne) [inline x/tokenomics/keeper/settle_session_accounting_test.go:214]: Same here.

Could you also please add a TODO in `settle_session_accounting.go` about checking that the application is staked for the session's service?

File: x/tokenomics/types/expected_keepers.go [hunk 4]
```diff
@@ -78,3 +78,10 @@ type SupplierKeeper interface {
 	GetSupplier(ctx context.Context, supplierAddr string) (supplier sharedtypes.Supplier, found bool)
 	SetSupplier(ctx context.Context, supplier sharedtypes.Supplier)
 }
+
+type ServiceKeeper interface {
+	GetService(ctx context.Context, serviceID string) (sharedtypes.Service, bool)
+	// NOTE: SetService is not used by the tokenomics keeper.
+	// It is only defined here to make it easier to add services to the service module in the tests.
+	SetService(ctx context.Context, service sharedtypes.Service)
```
Comment (red-0ne) [inline x/tokenomics/types/expected_keepers.go:86]: Please check if this is still needed when `SetService` is replaced by `WithService` in the tests.

**Comment by Olshansk** [comment]
These are all known E2E failures. @red-0ne and I are actively working on mitigating them:

Screenshot 2024-08-07 at 11 36 00 AM
Screenshot 2024-08-07 at 11 35 53 AM
Screenshot 2024-08-07 at 11 35 45 AM

**Review by red-0ne — COMMENTED** [review]
Great PR, had a couple for change requests, otherwise it's GTG

File: x/tokenomics/types/expected_keepers.go (hunk 4 above)
Comment (adshmh) [reply x/tokenomics/types/expected_keepers.go:86]: Thank you for the review. This is still needed: 

`TokenomicsModuleKeepersOpt` operates using `TokenomicsModuleKeepers`:

image


The `TokenomicsModuleKeepers` uses the `tokenomicstypes.ServiceKeeper` interface:

image

The `WithService`, as a `TokenomicsModuleKeepersOpt` can only access methods declared in the `expected_keepers.go`.

**Pull request merged**